### PR TITLE
Extract Search Client as a separate crate.

### PR DESCRIPTION
**Pull request by m-doughty**
# Extract Search Client as a separate crate

_Make the Search Client a reusable component which can be shared between Rust code bases which want to search the index._

relates to #401, #403, #620, #621.

### Acceptance Criteria

- [x] The Azure Search Client crate should be self-contained;
- [x] Azure Search Client code should be removed from the doc-index-updater sub-repo;
- [x] doc-index-updater should work as expected using the Azure Search Crate.

### Testing information

_notes that might be helpful for testing_

### Pre-flight Checklist

- [ ] Testing aligns with [testing strategy][testing strategy]
- [ ] DUXD review approval
- [ ] Accessibility Reviewed
- [ ] Product owner demo

### Pre-merge Checklist

- [ ] Branch test approved

[testing strategy]: https://github.com/MHRA/products/blob/master/docs/principles/testing.md "MHRA/products testing strategy"

### CODE REVIEW

File: medicines/search-client/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "search-client"
+version = "0.1.0"
+authors = ["Matt Doughty <matt.doughty@mhra.gov.uk>", "Craig Anderson <craig.anderson@red-badger.com>"]
```
Comment (majikandy) [inline medicines/search-client/Cargo.toml:4]: is this intentional?

Comment (m-doughty) [reply medicines/search-client/Cargo.toml:4]: Is what intentional?

Comment (majikandy) [reply medicines/search-client/Cargo.toml:4]: having specific developer names

Comment (m-doughty) [reply medicines/search-client/Cargo.toml:4]: I thought we had those in all of our Cargo.toml files?

Comment (craiga) [reply medicines/search-client/Cargo.toml:4]: You need an email address in your `Cargo.toml` in order to publish to crates.io.

File: .github/workflows/doc-index-updater-branch.yaml [hunk 2]
```diff
@@ -21,6 +21,10 @@ jobs:
       - name: Clone Repo
         uses: actions/checkout@v2
 
+      - name: Test for RC release
+        working-directory: ./medicines/doc-index-updater
+        run: cmd=`cat Cargo.toml | grep mhra_products_search_client | sed -e 's/"//' | sed -e 's/"//'` && if [[ $cmd == *"-rc"* ]];then echo rc dependencies not allowed for this package; exit 1; fi
```
Comment (majikandy) [inline .github/workflows/doc-index-updater-branch.yaml:26]: can be simplified to:-
`case $(grep <Cargo.toml mhra_products_search_client) in *"-rc"*) echo "Only stable versions of mhra_products_search_client should be used"; exit 1 ;; esac`

**Comment by StuartHarris** [comment]
Please see my other [comment](https://github.com/MHRA/products/pull/650#issuecomment-612076592)

**Comment by craiga** [comment]
> I'm not sure that we should be publishing like this to crates.io. As they stand, these crates don't seem to be generally useful right now (especially as they have mhra in the title) and so IMO don't belong there (yet).

@StuartHarris Here's a brief history of how we got here:

* @m-doughty, @majikandy and I moved the search client into its own crate in `medicines/search-client` so it could be shared between doc index updater and the API.
* We found that didn't work inside Docker containers, as Docker can only use files which are inside the directory the Dockerfile lives in.
* We talked about moving Dockerfiles about, changing build scripts, etc., but thought that pushing the shared package was a more elegant solution than polluting the monorepo with Dockerfiles a level about where they should be.

I agree that this isn't a great solution, so I'd be interested in hearing your thoughts on this. Do you think moving Dockerfiles would be a better solution than pushing this crate? Do you know of any other solutions to this problem?

I also agree that we shouldn't be pushing additional temp packages. I dropped the ball by not giving @majikandy permission to push to that crate before leaving for the long weekend.

**Comment by StuartHarris** [comment]
OK, thanks. We should definitely find another solution than pushing to crates.io. Those 3 crates will now be there forever, and that's not cool. We should only use crates.io for crates that we want to share with the world.
I would suggest either moving the `Dockerfile` up, or, even better, using `tar` to send the relevant context to the `docker build`. e.g. `tar cvf ../whatever | docker build -`

File: medicines/doc-index-updater/Cargo.toml [hunk 3]
```diff
@@ -20,6 +20,7 @@ md5 = "0.7.0"
 redis = { version = "0.15.1", features = ["tokio-rt-core"] }
 regex = "1.3.1"
 reqwest = { version = "0.10.4", features = ["json"] }
+search_client =  { git = "https://github.com/MHRA/products", branch = "createable-deleteable-into-search-crate" }
```
Comment (rjkerrison) [inline medicines/doc-index-updater/Cargo.toml:23]: I think this is the problem that [Cargo workspaces][1] solve. It seems weird for a repo to reference itself like this.

[1]: https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html

Comment (majikandy) [reply medicines/doc-index-updater/Cargo.toml]: Ok cool, we'll try it and compare

Comment (majikandy) [reply medicines/doc-index-updater/Cargo.toml]: the issue with workspaces is that you then reference the search-client package by doing `search_client = { path = ../search-client }` and then `make docker-build` won't work as the search-client is below the root of where the Dockerfile is.   So we could change all that too and all the build/release process but potentially a lot of changes (maybe too much right now) to achieve this?

Comment (majikandy) [reply medicines/doc-index-updater/Cargo.toml]: changed to use tar to pipe both the `doc-index-updater` and the `search-client` into `docker build` (via `make docker-build`)

**Comment by majikandy** [comment]
Now implemented using tar to pass doc-index-updater with dependency as a single tarball piped into docker build. No longer needs crates.io or GitHub url reference. Just a simple .. folder path

File: .github/workflows/search-client-master.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    steps:
+      - name: Clone Repo
+        uses: actions/checkout@v2
+        with:
+          path: products
+
+      - name: Test
+        working-directory: ./medicines/search-client
+        run: cargo test
```
Comment (sgrowe) [inline .github/workflows/search-client-master.yaml:24]: `search-client-branch.yaml` and `search-client-master.yaml` do the same thing so we could potentially just have one of them

File: medicines/doc-index-updater/Makefile [hunk 5]
```diff
@@ -32,13 +32,12 @@ ci-branch: docker-build ## Build branch in CI
 ci-master: ci-branch docker-push ## Build master in CI
 
 .PHONY: docker-build
-docker-build: ## Build and tag Docker image
-	docker build \
+docker-build: ## Build and tag Docker image (with search-client create as dependency)
+	tar cv --exclude=target Dockerfile -C .. ./doc-index-updater ./search-client | docker build - \
```
Comment (sgrowe) [inline medicines/doc-index-updater/Makefile:36]: Would it perhaps be simpler just to move the docker file up one folder (and maybe rename it to something like `Dockerfile.doc-index-updater`) so then you could just do `docker build -f Dockerfile.doc-index-updater`?

File: medicines/search-client/docs/development-and-releasing.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# Developing and Releasing this Package
+
+## Release Candidates
+
+To publish a new, release candidate version of this package:
+
+1. Set a new version number in `Cargo.toml` with an `-rc` suffix (for example, `1.2.2` is incremented to `1.2.3-rc1`, `1.2.3-rc4` is incremented to `1.2.3-rc5`).
```
Comment (sgrowe) [inline medicines/search-client/docs/development-and-releasing.md:7]: Is this file still needed if we're not pushing to crates.io anymore?

Comment (majikandy) [reply medicines/search-client/docs/development-and-releasing.md:7]: Nope, good point.

File: medicines/doc-index-updater/Makefile (hunk 5 above)
Comment (majikandy) [reply medicines/doc-index-updater/Makefile:36]: We did consider that but went with this so that the Dockerfile for doc-index-updater is actually with it - could get messy in the folder above if more appear over time.

File: .github/workflows/search-client-master.yaml (hunk 4 above)
Comment (majikandy) [reply .github/workflows/search-client-master.yaml:24]: True, just matches the pattern of the other packages and allows it to diverge.
